libraries/persistent_store/src/store.rs

```diff
@@ -1044,57 +1045,63 @@ impl<S: Storage> Store<S> {
     /// Parses the init info of a page.
     fn parse_init(&self, page: Nat) -> StoreResult<WordState<InitInfo>> {
         let index = self.format.index_init(page);
-        let word = self.storage_read_slice(index, self.format.word_size());
-        self.format.parse_init(Word::from_slice(word))
+        let mut word = WordSlice::default();
+        self.storage_read_slice(index, &mut word);
+        self.format.parse_init(Word::from_slice(&word))
     }
 
     /// Parses the compact info of a page.
     fn parse_compact(&self, page: Nat) -> StoreResult<WordState<CompactInfo>> {
         let index = self.format.index_compact(page);
-        let word = self.storage_read_slice(index, self.format.word_size());
-        self.format.parse_compact(Word::from_slice(word))
+        let mut word = WordSlice::default();
+        self.storage_read_slice(index, &mut word);
+        self.format.parse_compact(Word::from_slice(&word))
     }
 
     /// Parses a word from the virtual storage.
     fn parse_word(&self, pos: Position) -> StoreResult<WordState<ParsedWord>> {
         self.format
-            .parse_word(Word::from_slice(self.read_word(pos)))
+            .parse_word(Word::from_slice(&self.read_word(pos)))
     }
 
     /// Reads a slice from the virtual storage.
     ///
     /// The slice may span 2 pages.
     fn read_slice(&self, pos: Position, length: Nat) -> Vec<u8> {
-        let mut result = Vec::with_capacity(length as usize);
+        let mut result = vec![0; length as usize];
         let index = pos.index(&self.format);
         let max_length = self.format.page_size() - usize_to_nat(index.byte);
-        result.extend_from_slice(self.storage_read_slice(index, min(length, max_length)));
+        self.storage_read_slice(index, &mut result[..min(length, max_length) as usize]);
         if length > max_length {
             // The slice spans the next page.
             let index = pos.next_page(&self.format).index(&self.format);
-            result.extend_from_slice(self.storage_read_slice(index, length - max_length));
+            self.storage_read_slice(index, &mut result[max_length as usize..]);
```

Comment:
Double checking: I was wondering how much more copying we will have to do with this PR. Looks like we have to copy up to 2 pages now, instead of just the final slice that has always been a copy?

Reply:
Do you have a store benchmark you can quickly run, out of curiousity?

Reply:
Good point, I forgot about it. I've updated the numbers. Here's a summary:
- Boot is 35% slower
- Compaction is 5% slower
- Insert is 30% slower
- Remove is 35% slower

Reply:
Indeed, there's an alternative that we may prefer: #487 .

I believe there would be an ideal alternative. I don't think it's hard to implement, so I might just try.